File: packages/guards/src/indirect-injection.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface IndirectInjectionOptions {
  action: 'block' | 'warn';
  extraPatterns?: RegExp[];
}

const INDIRECT_PATTERNS: RegExp[] = [
  /\[INST\]/i,
  /\[\/INST\]/i,
  /<\|system\|>/i,
  /<\|user\|>/i,
  /<\|assistant\|>/i,
  /<<SYS>>/i,
  /\bHuman:\s/,
  /\bAssistant:\s/,
  /\bSystem:\s.*\b(ignore|override|forget|disregard)\b/i,
  /\bIMPORTANT:\s*(ignore|override|forget|new instructions)/i,
  /\bNEW INSTRUCTIONS:/i,
  /\bADMIN OVERRIDE:/i,
  /\bDEVELOPER MODE:/i,
  /```\s*(system|instructions?)\b/i,
  /\bignore\s+(the\s+)?(above|previous|prior)\s+(context|text|document)/i,
  /\binstead\s*,?\s*(respond|reply|answer|output)\s+with/i,
];

export function indirectInjection(options: IndirectInjectionOptions): Guard {
  const patterns = [...INDIRECT_PATTERNS, ...(options.extraPatterns ?? [])];

  return {
    name: 'indirect-injection',
    version: '0.1.0',
    description: 'Detects indirect prompt injection via RAG contexts, tool outputs, or embedded documents',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of patterns) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 4, 1.0) : 0;

      return {
        guardName: 'indirect-injection',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length, type: 'indirect-injection' } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/indirect_injection.py
"""Detect indirect prompt injection via RAG contexts and tool outputs."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_DEFAULT_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"\[INST\]", re.IGNORECASE),
    re.compile(r"<\|system\|>", re.IGNORECASE),
    re.compile(r"<\|user\|>", re.IGNORECASE),
    re.compile(
        r"IMPORTANT:\s*ignore",
        re.IGNORECASE,
    ),
    re.compile(r"ADMIN\s+OVERRIDE", re.IGNORECASE),
    re.compile(r"DEVELOPER\s+MODE", re.IGNORECASE),
    re.compile(
        r"ignore\s+the\s+above\s+context",
        re.IGNORECASE,
    ),
    re.compile(
        r"instead\s+respond\s+with",
        re.IGNORECASE,
    ),
    re.compile(
        r"<\|im_start\|>", re.IGNORECASE,
    ),
    re.compile(
        r"<<\s*SYS\s*>>", re.IGNORECASE,
    ),
]


class _IndirectInjection:
    def __init__(
        self,
        *,
        action: str = "block",
        extra_patterns: Optional[List[re.Pattern[str]]] = None,
    ) -> None:
        self.name = "indirect-injection"
        self.action = action
        self._patterns = list(_DEFAULT_PATTERNS)
        if extra_patterns:
            self._patterns.extend(extra_patterns)

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in self._patterns:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = min(len(matched) / 3, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="indirect-injection",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Indirect prompt injection detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "matched_patterns": len(matched),
                    "reason": (
                        "Text contains indirect injection"
                        " patterns embedded in context"
                        " or tool outputs"
                    ),
                }
                if triggered
                else None
            ),
        )


def indirect_injection(
    *,
    action: str = "block",
    extra_patterns: Optional[List[re.Pattern[str]]] = None,
) -> _IndirectInjection:
    return _IndirectInjection(
        action=action, extra_patterns=extra_patterns
    )
